### DISCUSSION TITLE
ECH (Encrypted ClientHello) support in Bromite

### DISCUSSION THREAD

**Discussion by eighthave** [bromite/bromite#2470] [General]
I just saw that Bromite has enabled ECH by default.  As part of the [DEfO](https://defo.ie) project, @sftcd @niallor and I are working on the ECH standard in the IETF as well as on implementing and spreading ECH across the internet.  We're interested in hearing the experiences of Bromite users and developers on how it is working.  Also, it would be useful for us to hear about what kind of planning or testing went into this decision so we can learn from the experience and help other projects with the transition.

**Comment by csagan5** [comment]
> I just saw that Bromite has enabled ECH by default. As part of the [DEfO](https://defo.ie) project, @sftcd @niallor and I are working on the ECH standard in the IETF as well as on implementing and spreading ECH across the internet.

Great to hear that you are involved there!

> We're interested in hearing the experiences of Bromite users and developers on how it is working.

It has been enabled for 23 days so far, I do not have figures about number of users but I expect that many Bromite users use DoH and I have not seen any bug report about it until now; if there was something wrong I believe it would already have popped up (AFAIK CloudFlare and other major players already support it).

> Also, it would be useful for us to hear about what kind of planning or testing went into this decision 

This is unfortunately not a strong suit of the project; what usually happens is that someone proposes an improvement (usually me or @uazo, but anyone can) related to privacy or security, we discuss it, we review the code and then enable it.
In this case I had taken note of the feature flag from Chromium's changelog some time ago, then before Bromite's v106 release I reviewed Chromium's implementation and it seemed fine to me (it has opportunistic support of ECH and fallback otherwise and no further privacy or security concerns found there); before release some manual testing is performed using devices and emulators.

> so we can learn from the experience and help other projects with the transition.

If I were to do this properly then the testing for ECH would include Chromium's own tests, plus testing with ECH-enabled webservers and also webservers behind a load-balancer that flap between ECH-enabled and not (to verify that no crashes happen and that session state is working as intended). I see you have some related resources on that site; perhaps publishing an official testsuite would help other projects.

**Reply by uazo** [reply]
> I just saw that Bromite has enabled ECH by default

Yeah, I didn't even notice that....
I think it's a good thing only because we disabled TLS resumption, otherwise is it possible that it is a new tracking mechanism?

>  I reviewed Chromium's implementation and it seemed fine to me 

correct me if i am wrong.
in chromium, the ech configuration is an array of bytes read from the https dns response and passed to boringssl opaquely, without verification.
boring reads it and saves it in the ssl configuration (which chromium persistently maintains for many, too many, days).
if the server rejects the first handshake, the connection can be retried with a configuration sent directly from the server (ref https://www.ietf.org/archive/id/draft-ietf-tls-esni-13.html#section-6.1.6).
here the specification says that it should not be saved for future connections, but in the chromium code I don't see this condition, can you help me you find it? I've follow `SSLClientSocketImpl::GetECHRetryConfigs()`
because otherwise is kept in the cache of the TLS resumption and reused in the future.

**Reply by csagan5** [reply]
> > I just saw that Bromite has enabled ECH by default
> 
> Yeah, I didn't even notice that....

It is in the changelog.

> I think it's a good thing only because we disabled TLS resumption, otherwise is it possible that it is a new tracking mechanism?

ECH is an improvement over ESNI; can you please explain how it is related to TLS resumption and how it could be a new tracking mechanism? The DoH traffic is out of band from the perspective of the server interacting with client during the ECH handshakes.

> in chromium, the ech configuration is an array of bytes read from the https dns response and passed to boringssl opaquely, without verification.

What type of verification should Chromium do in this case? I can only think of maximum size but I thought it's already hard capped when the ECH configuration is read from the response.

> boring reads it and saves it in the ssl configuration

It validates it via `ssl_is_valid_ech_config_list` first.

> which chromium persistently maintains

What do you mean by "persistently"?

> for many, too many, days

I do not follow, this cannot be used for fingerprinting. The server-side does not know whether client already had this configuration stored somewhere or not. What is the vector of abuse that you are thinking about?

> if the server rejects the first handshake, the connection can be retried with a configuration sent directly from the server (ref https://www.ietf.org/archive/id/draft-ietf-tls-esni-13.html#section-6.1.6).
> here the specification says that it should not be saved for future connections, but in the chromium code I don't see this condition, can you help me you find it? I've follow `SSLClientSocketImpl::GetECHRetryConfigs()`
> because otherwise is kept in the cache of the TLS resumption and reused in the future.

I don't see this. I see that for every `SSLClientSocketImpl::Connect()` a new SSL object is setup, which sets the ECH configuration. The ECH configuration received from server looks tied to the lifecycle of the SSL client socket.

I welcome scrutiny and review of any kind but please be more detailed and explicit!

**Reply by uazo** [reply]
you are absolutely right. I confused the two feature, I apologise

**Reply by csagan5** [reply]
> I apologise

?

On the contrary, if you have any question this would be a good time to ask.

**Comment by gothmog123** [comment]
Working great
